crates/ntx-builder/src/state/mod.rs

```diff
@@ -146,13 +180,42 @@ impl State {
             }
 
             self.in_progress.insert(candidate);
-            return TransactionCandidate { account: account.latest_account(), notes }.into();
+            return TransactionCandidate {
+                account: account.latest_account(),
+                notes,
+                chain_tip_header: self.chain_tip_header.clone(),
+                chain_mmr: self.chain_mmr.clone(),
+            }
+            .into();
         }
         self.inject_telemetry();
 
         None
     }
 
+    /// The latest block number the state knows of.
+    pub fn chain_tip(&self) -> BlockNumber {
+        self.chain_tip_header.block_num()
+    }
+
+    /// Updates the chain tip and MMR block count.
+    ///
+    /// Blocks in the MMR are pruned if the block count exceeds the maximum.
+    fn update_chain_tip(&mut self, tip: BlockHeader) {
+        // Update MMR which lags by one block.
+        self.chain_mmr.add_block(self.chain_tip_header.clone(), true);
+
+        // Set the new tip.
+        self.chain_tip_header = tip;
+
+        // Prune MMR if necessary.
+        if self.chain_mmr.num_tracked_blocks() > MAX_BLOCK_COUNT {
+            let pruned_block_height =
+                (self.chain_mmr.chain_length().as_usize() - PRUNED_BLOCK_COUNT) as u32;
+            self.chain_mmr.prune_to(..pruned_block_height.into());
+        }
```

Comment:
Is pruning multiple blocks cheaper than just pruning them as you go? As in, why not just `push` and `pop` to keep the amount retained constant?

Reply:
I can't see a way to push/pop via `PartialBlockchain` or `PartialMmr` atm. We could add it and update this. I'll proceed with current impl and we can follow up if need.

Reply:
I think my actual question is why not always prune to a constant length.

```suggestion
        self.chain_mmr.prune_to(..self.chain_mmr.chain_length()-PRUNED_BLOCK_COUNT)
```